ethereum/contracts/ShutdownSwitch.sol

```diff
@@ -0,0 +1,134 @@
+    function enabledFlag() public view returns (bool) {
+        return enabled;
+    }
+
+    /// @dev Event published whenenver a valid guardian votes.
+    event ShutdownVoteCast(address indexed voter, bool votedToEnable, uint16 numVotesToDisable, bool enableFlag);
+
+    /// @dev Event published whenever the shutdown status changes from enabled to disabled, or vice versa.
+    event ShutdownStatusChanged(bool enabledFlag, uint16 numVotesToDisable);
+
+    /// @dev Function that must be implemented by contracts inheriting from this one.
+    function getCurrentGuardianSet() public virtual view returns (Structs.GuardianSet memory);
+
+    /// @dev The number of disable votes required to disable transfers (assuming there are at least that many guardians).
+    uint16 constant REQUIRED_NO_VOTES = 3;
+
+    /// @dev Function that should be called from setup() on new contract deployments, or initialize() on the initial upgrade to deploy this feature.
+    function setUpShutdownSwitch() internal {
+        enabled = true;
+    }
+
+    /// @dev modifier used to block transfers when shutdown.
+    modifier isEnabled {
+        require(enabledFlag(), "transfers are temporarily disabled");
+        _;
+    }
+
+    /// @dev This is the function that allows guardians to vote, and determines the resulting shutdown status.
+    function castShutdownVote(bool _enabled) public {
+        // We always use the current guardian set.
+        Structs.GuardianSet memory gs = getCurrentGuardianSet();
+
+        // Only currently active guardians are allowed to vote.
+        require(isRegisteredVoter(msg.sender, gs), "you are not a registered voter");
+
+        // Only votes to disable are maintained in the map.
+        if (_enabled) {
+            delete disabledVotes[msg.sender];
+        } else {
+            disabledVotes[msg.sender] = true;
+        }
+
+        // Update the number of votes to disable.
+        uint16 votesToDisable = computeNumVotesDisabled(gs);
+
+        // Determine the new shutdown status and generate the appropriate events.
+        bool newEnabledFlag = (votesToDisable < computeRequiredVotesToDisable(gs.keys.length));
+        emit ShutdownSwitch.ShutdownVoteCast(msg.sender, _enabled, votesToDisable, newEnabledFlag);
+
+        if (newEnabledFlag != enabled) {
+            enabled = newEnabledFlag;
+            emit ShutdownSwitch.ShutdownStatusChanged(newEnabledFlag, votesToDisable);
+        }
+    }
+
+    /// @dev Determines if a voter is a current guardian.
+    function isRegisteredVoter(address voter, Structs.GuardianSet memory gs) private pure returns(bool) {
+        for (uint idx = 0; (idx < gs.keys.length); idx++) {
+            if (voter == gs.keys[idx]) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /// @dev Returns the number of disable votes required to go disabled.
+    function computeRequiredVotesToDisable(uint numGuardians) private pure returns(uint16) {
```

Comment:
can we call this `requiredVotesToShutdown` or something like that? 'disable' is a little confusing to me, because it's not immediately clear if it disables the shutdown, or transfers (I know it's the latter after reading the code)

Reply:
I agree, I think this would read cleaner.

Reply:
Renamed things to use Shutdown rather than Disable.